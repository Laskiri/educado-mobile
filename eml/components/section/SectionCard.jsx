import React from 'react';
import { View, TouchableOpacity } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import Text from '../general/Text';
import PropTypes from 'prop-types';


/**
 * A component that displays a section card with collapsible content.
 * @param {Object} section - The section object containing the section data.
 * @param {Number} progress - The progress containing the student's progress.
 * @param {Function} onPress - The callback function to navigate
 * @returns {JSX.Element} - The SectionCard component.
 */
export default function SectionCard({ section, progress, onPress }) {


	const isComplete = progress === section.components.length;
	const inProgress = 0 < progress && progress < section.components.length;
	const progressText = isComplete ? 'Concluído' : inProgress ? 'Em progresso' : 'Não iniciado';
	const progressTextColor = isComplete ? 'text-success' : 'text-projectBlack';


	return (
		<View>
			<TouchableOpacity
				className="bg-secondary border-[1px] border-lightGray rounded-lg shadow-lg shadow-opacity-[0.3] mb-[15] mx-[18] overflow-hidden elevation-[8]"
				onPress={onPress}
			>
				<View className="flex-row items-center justify-between px-[25] py-[15]">
					<View>
						<Text className="text-[16px] font-bold text-projectBlack flex-[1]">
							{section.title}
						</Text>
						<Text className={`mr-[10] ${progressTextColor}`}>
							{/* progress */}
							{progress}/{section.components.length} {progressText}
							{(isComplete ?
								<MaterialCommunityIcons
									testID={'check-circle'}
									name={'check-circle'}
									size={16}
									color="green"
								/> : ''
							)}
						</Text>
					</View>
					<MaterialCommunityIcons
						testID="chevron-right"
						name="chevron-right"
						size={25}
						color="gray"
					/>
				</View>
			</TouchableOpacity>
			
		</View>
	);
}

SectionCard.propTypes = {
	section: PropTypes.object,
	progress: PropTypes.number,
	onPress: PropTypes.func,
};