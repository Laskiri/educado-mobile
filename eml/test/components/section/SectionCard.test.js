/**
 * This file contains the test suite for the SectionCard component.
 * It imports React, render, fireEvent and renderer from testing-library/react-native.
 * It also imports the SectionCard component from the components/section directory.
 * The useNavigation hook is mocked using jest.mock.
 * A sample section object is created for testing purposes.
 * The test suite includes tests for rendering the component with provided data,
 * displaying the correct status based on progress, and expanding and collapsing the component.
 * @module SectionCard.test
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import SectionCard from '../../../components/section/SectionCard';
import { mockDataAsyncStorage } from '../../mockData/mockDataAsyncStorage';

// Mock the useNavigation hook outside the describe block
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

let sectionCard;

// Sample section data for testing
const mockData = mockDataAsyncStorage();

beforeEach(async () => {
  sectionCard = renderer.create(<SectionCard section={mockData.section} />);
});

afterAll(() => {
  jest.resetModules();
  jest.restoreAllMocks();
});

describe('<SectionCard />', () => {
  it('Renders correctly with provided data', async () => {
    const { getByText } = render(<SectionCard section={mockData.section} />);

    // Check if title and description are displayed
    expect(getByText(mockData.section.title)).toBeTruthy();
  });

  /*it('Displays correct status based on progress', async () => {
    const { queryByText } = render(<SectionCard section={mockData.section} />);

    // Use a regular expression to match the 0/total pattern
    const pattern = new RegExp(`0/${mockData.section.total}`);
    expect(queryByText(pattern)).toBeTruthy();
  });*/
  /**
    * Tests if the SectionCard component renders correctly.
    */
  it('renders SectionCard correctly', async () => {
    expect(await sectionCard.toJSON()).toMatchSnapshot();
  });

  it('should trigger onPress when the card is pressed', () => {
    // Mock data for testing
    const mockData = {
      section: {
        title: 'Sample Section',
        components: ['Component 1', 'Component 2', 'Component 3'],
      },
      progress: 1, // Example progress value
    };

    // Mock the onPress function
    const mockOnPress = jest.fn();

    // Render the SectionCard component
    const { getByText, getByTestId } = render(
      <SectionCard
        section={mockData.section}
        progress={mockData.progress}
        onPress={mockOnPress}
      />
    );

    // Locate the TouchableOpacity by the section title text
    const touchableElement = getByText(mockData.section.title);

    // Simulate a press event
    fireEvent.press(touchableElement);

    // Assert that the onPress function was called
    expect(mockOnPress).toHaveBeenCalled();
  });
});