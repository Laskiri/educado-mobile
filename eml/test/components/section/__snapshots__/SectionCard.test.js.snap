// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SectionCard /> renders SectionCard correctly 1`] = `
<View>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#F1F9FB",
        "borderBottomColor": "#E4E4E4",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderBottomWidth": 1,
        "borderLeftColor": "#E4E4E4",
        "borderLeftWidth": 1,
        "borderRightColor": "#E4E4E4",
        "borderRightWidth": 1,
        "borderTopColor": "#E4E4E4",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderTopWidth": 1,
        "elevation": 8,
        "marginBottom": 15,
        "marginLeft": 18,
        "marginRight": 18,
        "opacity": 1,
        "overflow": "hidden",
        "shadowColor": "rgba(0, 0, 0, 0.1)",
        "shadowOffset": {
          "height": 10,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 15,
      }
    }
  >
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "center",
          },
          {
            "justifyContent": "space-between",
          },
          {
            "paddingLeft": 25,
            "paddingRight": 25,
          },
          {
            "paddingBottom": 15,
            "paddingTop": 15,
          },
        ]
      }
    >
      <View>
        <Text
          style={
            [
              {
                "fontFamily": "Montserrat-Regular",
              },
              {
                "color": "#383838",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "fontSize": 16,
                },
                {
                  "fontFamily": "Montserrat-Bold",
                },
                {
                  "color": "#383838",
                },
                {
                  "marginBottom": 8,
                },
              ],
            ]
          }
        >
          test section
        </Text>
        <View
          style={
            [
              {
                "flexDirection": "row",
              },
              {
                "alignItems": "center",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "fontFamily": "Montserrat-Regular",
                },
                {
                  "color": "#383838",
                },
                {
                  "fontSize": 16,
                },
                [
                  {
                    "fontSize": 14,
                  },
                  {
                    "fontFamily": "Montserrat-Regular",
                  },
                  {
                    "color": "#383838",
                  },
                ],
              ]
            }
          >
            /
            0
             
            Não iniciado
          </Text>
          <View
            style={
              [
                {
                  "marginLeft": 8,
                },
              ]
            }
          />
        </View>
      </View>
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "gray",
              "fontSize": 25,
            },
            undefined,
            {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
        testID="chevron-right"
      >
        󰅂
      </Text>
    </View>
  </View>
</View>
`;
